pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -787,26 +788,81 @@ func (a Adapter) ExtractProjectToComponent(componentInfo common.ComponentInfo, t
+	deployCmd, err := getDeployCommand(a)
+	if err != nil {
+		return err
+	}
+
+	return a.ExecuteDevfileCommand(deployCmd, true)
+}
+
+// UnDeploy reverses the effect of the 'deploy' command defined in a devfile
+func (a Adapter) UnDeploy() error {
+	deployCmd, err := getDeployCommand(a)
+	if err != nil {
+		return err
+	}
+	// Only get commands of type 'Apply'; those are the only ones we're concerned with for 'Deploy'
+	// 'Deploy' command does not support any other command type
+	allApplyCommands, err := a.Devfile.Data.GetCommands(parsercommon.DevfileOptions{
 		CommandOptions: parsercommon.CommandOptions{
-			CommandGroupKind: devfilev1.DeployCommandGroupKind,
+			CommandType: devfilev1.ApplyCommandType,
 		},
 	})
 	if err != nil {
-		return nil
+		return err
 	}
 
-	if len(commands) == 0 {
-		return errors.New("error deploying, no default deploy command found in devfile")
+	// Get components of type Kubernetes; one of these components should be associated with one of the 'Deploy' composite commands
+	kubeComponents, err := a.Devfile.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: devfilev1.KubernetesComponentType},
+	})
+	if err != nil {
+		return err
 	}
 
-	if len(commands) > 1 {
-		return errors.New("more than one default deploy command found in devfile, should not happen")
+	// A list of Command objects of the commands that form the composite 'Deploy' command
+	var requiredCommands []devfilev1.Command
+
+	// Get the Command object of commands that form the composite 'Deploy' command
+	for _, commandId := range deployCmd.Composite.Commands {
+		for _, cmd := range allApplyCommands {
+			if cmd.Id == commandId {
+				requiredCommands = append(requiredCommands, cmd)
+			}
+		}
 	}
 
-	deployCmd := commands[0]
+	// The component of type Kubernetes associated with one of the composite 'Deploy' commands
+	var component devfilev1.Component
 
-	return a.ExecuteDevfileCommand(deployCmd, true)
+	// Get the component of the command that deploys the k8s object
+	// At this point we assume the component list is of type Kubernetes, and the commands are that of the composite 'Deploy' command
+	for _, cmp := range kubeComponents {
+		for _, cmd := range requiredCommands {
+			if cmd.Apply.Component == cmp.Name {
+				component = cmp
+				break
+			}
+		}
+	}
+
+	// Parse the component's Kubernetes manifest
+	u, err := service.GetK8sComponentAsUnstructured(component.Kubernetes, a.Context, devfilefs.DefaultFs{})
+	if err != nil {
+		return err
+	}
+
+	kClient := a.Client.GetKubeClient()
+	// Get the REST mappings
+	gvr, err := kClient.GetRestMappingFromUnstructured(u)
+	if err != nil {
+		return err
+	}
+	log.Infof("\nUn-deploying the Kubernetes %s: %s", u.GetKind(), u.GetName())
+	// Un-deploy the K8s manifest
+	err = kClient.DeleteDynamicResource(u.GetName(), gvr.Resource.Group, gvr.Resource.Version, gvr.Resource.Resource)
+	return err
```

Comment:
this should be done in `pkg/devfile/adapters/kubernetes/component/component_kubernetes.go`